Project Duru

FabianFalck opened the pull request:
*Issue #, if available:*

None.

*Description of changes:*

Push current state of Project Duru.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Yes.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

FabianFalck committed:
Project Duru

FabianFalck committed:
reformatted with black

jaheba commented on src/gluonts/nursery/duru/duru/hvae.py:
```diff
@@ -0,0 +1,110 @@
+
+
+class HVAE(nn.Module):
+    def __init__(self):
+        super().__init__()
+        pass
```
```suggestion
```

If this is doing nothing, we can remove it.

FabianFalck replied:
I've removed the `pass`, it must call the super constructor

FabianFalck replied:
After discussion: Agreed, took out the superfluous `__init__` call.

jaheba commented on src/gluonts/nursery/duru/duru/hvae.py:
```diff
@@ -0,0 +1,110 @@
+
+
+class DecoderBlock(nn.Module):
+    def __init__(self):
+        super().__init__()
```
Same here.

FabianFalck replied:
again, must call super constructor

FabianFalck replied:
After discussion: Agreed, took out the superfluous `__init__` call.

jaheba commented on src/gluonts/nursery/duru/duru/hvae.py:
```diff
@@ -0,0 +1,110 @@
+                    self.H.z_channels,
+                    self.H.context_length + self.H.forecast_length,
+                )
+            )[i]
```
This looks weird. We are constructing a 3d zeros, to then just slice it directly?

FabianFalck replied:
yes, to get a list. probably better ways to do it. this one is not a very important class though.

FabianFalck committed:
reformatting again with correct version of black

jaheba commented on src/gluonts/nursery/duru/duru/hvae.py:
```diff
@@ -0,0 +1,110 @@
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+import torch as T
```
Is this common? Can't we just do `import torch`?

FabianFalck replied:
I agree, just changed it all to `torch`, also in the calls.

jaheba commented on src/gluonts/nursery/duru/duru/vdvae_conv.py:
```diff
@@ -0,0 +1,950 @@
+
+        self.layer_index_to_out_res = (
+            {}
+        )  # mapping from layer index to resolution of the output of that layer
```
Let's have these comments above the line they are describing to avoid these breaks.

FabianFalck replied:
done

jaheba commented on src/gluonts/nursery/duru/hyperparams.py:
```diff
@@ -0,0 +1,623 @@
+
+
+class Hyperparams(dict):
+    """
+    Wrapper class for a dictionary required for pickling.
+    """
```
Not sure I understand the need for this.

FabianFalck replied:
It's required in order to pickle it correctly. There was an issue that I don't fully recall if we don't do this, probably due to the diversity of types the dictionary is pointing to.

jaheba replied:
If you know the types of the variables, it's probably easier to use something like `pydantic` to define which fields are there.

FabianFalck replied:
this is nice. For simplicity, I would suggest to leave it as is for now

jaheba replied:
There is also types.SimpleNamespace which should also do the trick.

jaheba commented on src/gluonts/nursery/duru/README.md:
````diff
@@ -0,0 +1,64 @@
+Please set up a virtual environment installing the following key dependencies: 
+
+```shell
+pip install torch torchvision
````
You could add a requirements file for this.

FabianFalck replied:
Done.

FabianFalck committed:
License header corrections

FabianFalck committed:
Code review discussion - items addressed

Addressed all items in code review discussion.
Added conda environment versions and dependencies, also referred to it in README.

jaheba commented on src/gluonts/nursery/duru/__init__.py:
```diff
@@ -0,0 +1,12 @@
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
```
You probably need to run `black` again, since it requires a new line at the end of the file

FabianFalck replied:
Reran black.

FabianFalck replied:
Resolved now.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+# permissions and limitations under the License.
+import gluonts.dataset
+import torch
+import torch as T
```
Just `torch` over `T`?

FabianFalck replied:
took `import torch as T` out entirely as per your other comment, it wasn't used anyway. thank you!

jaheba replied:
Looks like it's still there?

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+import gluonts.dataset
```
Can you group the imports by

* Standard library imports.
* Related third party imports.
* Local application/library specific imports.

See: https://peps.python.org/pep-0008/#imports

FabianFalck replied:
Done.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+
+    elif H.dataset in list(
+        dataset_recipes.keys()
+    ):  # all available datasets in gluonts: ['constant', 'exchange_rate', 'solar-energy', 'electricity', 'traffic', 'exchange_rate_nips', 'electricity_nips', 'traffic_nips', 'solar_nips', 'wiki-rolling_nips', 'taxi_30min', 'kaggle_web_traffic_with_missing', 'kaggle_web_traffic_without_missing', 'kaggle_web_traffic_weekly', 'm1_yearly', 'm1_quarterly', 'm1_monthly', 'nn5_daily_with_missing', 'nn5 ...
```
I think we can remove that comment. Available datasets might change in the feature.

FabianFalck replied:
I think it's really useful for the developer to know what is roughly behind dataset_receipes, because it's otherwise not clear from context. So I would rather leave it.

jaheba replied:
If I was interested in knowing what datasets are available, I can look them up myself. Also, we might add or remove datasets, so this would become quickly out of sync.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+        # X_test = torch.ones(n_samples_test, H.n_meas, n_steps).float().numpy()  # Conv1D expects (N,C_in,L_in), where C_in=number of channels=measurement dimension, L_in=length of signal sequence
+        # Y_test = torch.ones(n_samples_test).float().numpy()
+
+    elif H.dataset in list(
```
Double no! To check whether a key exist in a dictionary, you can just do `key in dct`. There is no need to do `key in dct.keys()`.

But never call `list` on it, since that turns the lookup from `O(1)` to `O(n)`.

FabianFalck replied:
Thanks! Done.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+from gluonts.dataset.rolling_dataset import generate_rolling_dataset
+
+
+def set_up_data(H):
```
What's `H`? I generally have an aversion against single-letter variable names.

FabianFalck replied:
H is the parsed command line object which contains all hyperparameters. I find the single-letter variable name convenient here since it's abundantly used and hence it should be clear what it means.

jaheba replied:
It sure sounds convenient, but for someone new to the code it is not obvious. Code is read more often then written, so I will insist on changing this to something more self explainatory.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+from gluonts.dataset.repository.datasets import get_dataset, dataset_recipes
+from gluonts.dataset.split import OffsetSplitter
+from utils import calculate_padding_base_2
+from gluonts.dataset.rolling_dataset import StepStrategy, NumSplitsStrategy
+from gluonts.dataset.rolling_dataset import generate_rolling_dataset
```
We've just removed this from `gluonts`.

Here is a notebook for the new way: https://ts.gluon.ai/dev/tutorials/data_manipulation/dataset_splitting_example.html

FabianFalck replied:
Thanks for this useful comment! I'm happy to change it in a future version, but would for now leave the code as is, using an older version of gluonts.

jaheba replied:
How would that work? Code in GluonTS can't depend on an older version of itself.

You would need to make `duru` it's on package then.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+import torchvision.transforms as transforms
+from gluonts.dataset.repository.datasets import get_dataset, dataset_recipes
+from gluonts.dataset.split import OffsetSplitter
+from utils import calculate_padding_base_2
```
This should be a relative import, `from .utils import ...`.

FabianFalck replied:
It works I guess, given that the project folder is the repository. Do we need to change it?

jaheba replied:
Yes!

This won't work if it's used as a package.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+
+    elif H.dataset == "sine":
+
+        # OLD -----
```
Can we remove this out-commented code?

FabianFalck replied:
done.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+        # OLD -----
+
+        # NEW 1: different noise, different starting point -----
+        class SineWaveGluonTSDataset(gluonts.dataset.Dataset):
```
I think we had issues with deriving from `Dataset` in the past. The idea is that it's a protocol (i.e. nominal typing), meaning that you don't need to derive from it.

FabianFalck replied:
it works right now, so I'd rather like to leave it. It's purpose is to overwrite and expect the __iter__ and __len__ method.

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+                        )
+                        noise = noise.astype(np.float32)
+
+                        item = {}
+                        item["start"] = 0.0  # dummy data
+                        item["target"] = noise
+                        item["item_id"] = str(t)
```
You can do this all in one statement.

FabianFalck replied:
I made this change

jaheba commented on src/gluonts/nursery/duru/data.py:
```diff
@@ -0,0 +1,488 @@
+                        self.data.append(item)
+
+            def __iter__(self):
+                for t in range(self.n_items):
+                    # print(t)
+                    yield self.data[t]
+                # raise StopIteration
```
Can't this just be
```suggestion
                yield from self.data
```

FabianFalck replied:
perhaps yes, I haven't tested it, but would leave it for now

jaheba replied:
Then let's just change this.

jaheba commented on src/gluonts/nursery/duru/duru.yml:
```diff
@@ -0,0 +1,111 @@
+name: duru
```
What is this file used for?

FabianFalck replied:
Storing logger details.

jaheba commented on src/gluonts/nursery/duru/vdvae_LICENSE:
```diff
@@ -0,0 +1,9 @@
+License of [VDVAE](https://github.com/openai/vdvae/blob/main/LICENSE.md).
```
Do we need to include this file? Can't we just link to it somewhere?

FabianFalck replied:
Yes, we are using VDVAE's code partly. I'm referring to this also in the appendix in the functions which we are using.

FabianFalck replied:
The MIT license of the repo states: 

"
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

***The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.***
"

Hence, I suggest to leave it in.

jaheba replied:
Can we include the license in the head of the files where it applies and delete the separate file?

jaheba commented on src/gluonts/nursery/duru/utils.py:
```diff
@@ -0,0 +1,423 @@
+import numpy as np
+
+
+def str2bool(v):
```
This looks very much like `distutils.util.strtobool`. Can't we just use that one?

FabianFalck replied:
Thanks for this great point! I would like to leave this function in as there seem to be problems with boolean parsing in command line arguments that I otherwise appear. See https://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse. I also tested this before.

FabianFalck committed:
reformatting again.

FabianFalck committed:
one statement change

FabianFalck committed:
removed commented out code

FabianFalck committed:
key lookup made more efficient

FabianFalck committed:
regrouping of imports

FabianFalck committed:
black reformatting of data.py

FabianFalck commented:
Hi @jaheba, 
I made a pass through all your remaining comments. I resolved some of them, on some I suggest to leave the code as is for now in the interest of not breaking things. I suggest you go over it once more, make the final decision, and I then resolve any remaining TODOs. 
Many thanks, 
Fabian

jaheba reviewed (commented):
It looks like mentioned code changes were not actually pushed.
